src/mod_cache_users.erl

```diff
@@ -0,0 +1,237 @@
+    del_domain(Tab, Domain, ets:first(Tab)),
+    Acc.
+
+%%====================================================================
+%% Helpers
+%%====================================================================
+
+-spec does_stored_user_exist(mongooseim:host_type(), jid:jid()) -> boolean().
+does_stored_user_exist(HostType, JID) ->
+    true =:= ejabberd_auth:does_stored_user_exist(HostType, JID).
+
+-spec does_cached_user_exist(mongooseim:host_type(), jid:lserver(), jid:luser()) -> boolean().
+does_cached_user_exist(HostType, LServer, LUser) ->
+    Key = key(LUser, LServer),
+    Tab = tbl_name(HostType),
+    is_member(Tab, Key, ets:first(Tab)).
+
+-spec put_user_into_cache(mongooseim:host_type(), jid:lserver(), jid:luser()) -> ok.
+put_user_into_cache(HostType, LServer, LUser) ->
+    Key = key(LUser, LServer),
+    Tab = tbl_name(HostType),
+    put_member(Tab, Key, ets:first(Tab)).
+
+-spec key(jid:luser(), jid:lserver()) -> {jid:lserver(), jid:luser()}.
+key(LUser, LServer) ->
+    {LServer, LUser}.
+
+is_member(_, _, '$end_of_table') ->
+    false;
+is_member(Tab, Key, Table) ->
+    case ets:lookup(Tab, Table) of
+        [{_, Ets}] ->
+            case ets:member(Ets, Key) of
+                false -> is_member(Tab, Key, ets:next(Tab, Table));
+                true -> true
+            end;
+        [] ->
+           is_member(Tab, Key, ets:next(Tab, Table))
+    end.
+
+%% There's a chance that by the time we insert in the ets table, this table is not
+%% the first anymore because the cleaner has taken action and pushed it behind.
+%% That's fine, worst case this record will live a segment less than expected.
+put_member(Tab, Key, Table) ->
+    Ets = ets:lookup_element(Tab, Table, 2),
+    ets:insert(Ets, {Key}),
+    ok.
+
+del_member(_, _, '$end_of_table') ->
+    ok;
+del_member(Tab, Key, Table) ->
+    case ets:lookup(Tab, Table) of
+        [{_, Ets}] ->
+            ets:delete(Ets, Key),
+            del_member(Tab, Key, ets:next(Tab, Table));
+        [] ->
+            del_member(Tab, Key, ets:next(Tab, Table))
+    end.
+
+del_domain(_, _, '$end_of_table') ->
+    ok;
+del_domain(Tab, Domain, Table) ->
+    case ets:lookup(Tab, Table) of
+        [{_, Ets}] ->
+            ets:match_delete(Ets, {{Domain, '_'}}),
+            del_domain(Tab, Domain, ets:next(Tab, Table));
+        [] ->
+            del_domain(Tab, Domain, ets:next(Tab, Table))
+    end.
+
+-spec start_cache(mongooseim:host_type(), gen_mod:module_opts()) -> any().
+start_cache(HostType, Opts) ->
+    Tab = tbl_name(HostType),
+    N   = gen_mod:get_opt(n, Opts, 1),
+    TTL = gen_mod:get_opt(ttl, Opts, infinity),
+    RC  = gen_mod:get_opt(read_concurrency, Opts, true),
+    WC  = gen_mod:get_opt(write_concurrency, Opts, true),
+    ParentTab = ets:new(Tab, [ordered_set, named_table, public, {read_concurrency, true}]),
+    lists:foreach(
+      fun(I) ->
+              Ref = ets:new(Tab, [set, public, {read_concurrency, RC}, {write_concurrency, WC}]),
```

Comment:
maybe just use named tables, like `Tab ++ SegmentNumStr`?

Also, if you check logs carefully, it would complain that you are creating ets tables from inside of `gen_mod:start_module`, which is big no. We can't guarantee that the caller of the callback would stay alive. Basically, it stops working with RPC-s. So, create the tables from the inside of the gen_server.

Reply:
Don't want to use named tables, I want the parent table to be the interface to the segments.

Creating the table from gen_mod, I think I know what to do there.